hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/ReadableKVStateBase.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.spi.state;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.*;
+
+/**
+ * A base class for implementations of {@link ReadableKVState} and {@link WritableKVState}.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+public abstract class ReadableKVStateBase<K extends Comparable<K>, V>
+        implements ReadableKVState<K, V> {
+    /** The state key, which cannot be null */
+    private final String stateKey;
+
+    /**
+     * A cache of all values read from this {@link ReadableKVState}. If the same value is read
+     * twice, rather than going to the underlying merkle data structures to read the data a second
+     * time, we simply return it from the cache. We also keep track of all keys read, which is
+     * critical for dealing with validating what we read during pre-handle with what may have
+     * changed before we got to handle transaction. If the value is "null", this means it was NOT
+     * FOUND when we looked it up.
+     */
+    private final Map<K, V> readCache = new HashMap<>();
```

Comment:
The [current `PreHandleWorkflow`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java#L102) will seemingly share the same `readCache` among multiple threads (each pre-handling a different transaction in an event using the same state).

Not necessarily an issue with this PR, just a comment.